### PR TITLE
[TFA] [osd ok-to-upgrade] stabilise the upgrade suite for osd ok-to-upgrade holistic upgrade tests

### PR DESCRIPTION
**Pull request by s-vipin**
- stabilise the upgrade suite for osd ok-to-upgrade holistic upgrade tests
- Addresses tfa 
    https://ibm-ceph.atlassian.net/browse/IBMCEPHQE-24433
    https://ibm-ceph.atlassian.net/browse/IBMCEPHQE-24434
    https://ibm-ceph.atlassian.net/browse/IBMCEPHQE-24435
- Adds negative and positive test cases for upgrade using crush buckets
          - bucket_type_bucket_name_rack
          - bucket_type_bucket_name_chassis
          - bucket_type_bucket_name_host
          - negative_bucket_type_bucket_name_rack_without_daemon_types_arg
          - negative_bucket_type_bucket_name_rack_host_arg
          - negative_bucket_type_bucket_name_services_arg
          - negative_bucket_type_bucket_name_topological_labels_arg
          
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by s-vipin** [comment]
Pass logs:- http://magna002.ceph.redhat.com/cephci-jenkins/vipin-runs/tfa-fixes-ok-to-upgrade/

File: tests/rados/test_osd_ok_to_upgrade_holistic_tests.py [hunk 1]
```diff
@@ -81,6 +89,10 @@ def run(ceph_cluster, **kw):
     # as mgr,mon,crash are already upgraded to latest version 9.1
     new_image = get_daemon_container_images(rados_obj, daemon="mgr")[0]
 
+    license_cmd = f"ceph orch accept-license --image {new_image}"
+    out, _ = cephadm_obj.shell(args=[license_cmd], check_status=False)
```
Comment (s-vipin) [inline tests/rados/test_osd_ok_to_upgrade_holistic_tests.py]: remove check_status=False

File: tests/rados/test_osd_ok_to_upgrade_holistic_tests.py [hunk 2]
```diff
@@ -428,27 +517,18 @@ def run(ceph_cluster, **kw):
         # Delete pools created for IO during upgrade
         rados_obj.delete_pool(pool=re_pool_name)
         log.info("Deleted pool: %s", re_pool_name)
+        mon_config_method.remove_config(
+            section="mgr", name="debug_mgr", verify_rm=False
```
Comment (s-vipin) [inline tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:532]: duplicate

**Comment by s-vipin** [comment]
# Detailed Code Review: tfa-fixes-ok-to-upgrade-holistic Branch

## Summary
Reviewed commit comparing `tfa-fixes-ok-to-upgrade-holistic` against `main` branch. The changes enhance OSD upgrade testing with CRUSH bucket support and improve test reliability.

## Files Changed
1. [`suites/tentacle/rados/tier-2_rados_test_ok_to_upgrade_holistic.yaml`](suites/tentacle/rados/tier-2_rados_test_ok_to_upgrade_holistic.yaml:346)
2. [`tests/rados/test_osd_ok_to_upgrade_holistic_tests.py`](tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:1)

## Key Changes

### Positive Improvements ✅
- **Fixed typo**: `host_wihtout_osd` → `host_without_osd`
- **Enhanced test coverage**: Added 7 new test scenarios for CRUSH bucket-based upgrades (rack, chassis, host)
- **Better error handling**: Added `check_ec=True` for proper command failure detection
- **Improved validation**: Restructured logic to properly validate ok-to-stop vs ok-to-upgrade usage
- **Correct CLI flags**: Changed to `--crush-bucket-type` and `--crush-bucket-name` (aligns with Ceph CLI)
- **Enhanced logging**: More precise grep patterns with `check_ec=False` to prevent false failures

### Areas of Concern ⚠️

**High Severity:**
1. **Complex downgrade function** (lines 702-790): New 88-line `_downgrade_osds_to_old_image()` function removes and re-adds OSDs individually, significantly increasing test time and complexity
2. **Validation logic improvement**: Now properly asserts mutual exclusivity of ok-to-stop and ok-to-upgrade commands

**Medium Severity:**
3. **Sequential OSD operations**: OSDs removed one-by-one with full PG cleanup between each, causing extended test duration
4. **DRY violation**: Four negative test scenarios (lines 306-393) repeat identical error-checking pattern
5. **Function length**: Downgrade function handles multiple responsibilities (config, removal, tracking, re-addition, service management)
6. **Race condition risk**: Service management state changes without explicit success verification
7. **Breaking change**: Final upgrade test configuration changed from verification-based to command-based approach

**Low Severity:**
8. **Missing documentation**: New scenarios not documented in module docstring
9. **Unused variable**: `cephadm_obj` created but never used
10. **Debug logging**: MGR debug set to 20/20 during tests (performance impact?)
11. **Command injection risk**: F-string command construction (mitigated by test-only context)

## Recommendations
1. Consider refactoring downgrade function into smaller, focused functions
2. Add helper function for repeated negative scenario validation
3. Update module docstring with new test scenarios
4. Remove unused `cephadm_obj` variable
5. Add explicit error handling in service state transitions
6. Document the breaking change in upgrade test configuration

**All 17 findings added to Bob Findings panel for detailed review.**

**Comment by s-vipin** [comment]
## Commit Review Rating: **7/10** ⭐⭐⭐⭐⭐⭐⭐

### Summary
Your commit "stabilise the upgrade suite for osd ok-to-upgrade tests" makes significant improvements to the OSD upgrade testing infrastructure. The changes add important test coverage and fix critical issues, but there are areas that need attention.

### ✅ Strengths

1. **Critical Bug Fix**: Fixed typo `host_wihtout_osd_obj` → `host_without_osd_obj`
2. **Enhanced Test Coverage**: Added 6 new test scenarios including negative test cases for bucket-type upgrades
3. **Improved Validation**: Better log validation for `ok-to-stop` vs `ok-to-upgrade` commands
4. **Robust Cleanup**: New `_downgrade_osds_to_old_image()` function provides proper test cleanup
5. **Better Debugging**: Added debug logging and log rotation for troubleshooting
6. **Type Safety**: Added type validation for metadata responses

### ⚠️ Areas for Improvement

**High Priority:**
- **Function complexity** (line 61-524): The `run()` function is 600+ lines - needs refactoring into smaller functions
- **Error handling** (line 737-740): Type validation raises ValueError but isn't caught by caller

**Medium Priority:**
- **Code duplication**: 4 negative test scenarios repeat identical patterns (lines 308-388)
- **Race conditions**: OSD downgrade logic could benefit from retry mechanisms
- **Log validation**: grep operations don't distinguish between "no matches" vs "command failed"

**Low Priority:**
- Magic numbers (timeout=12000) need named constants
- Inconsistent comment formatting throughout
- Log rotation happens inside loop unnecessarily

### 📊 Detailed Breakdown

- **Functionality**: 8/10 - Solid feature additions, but error handling gaps
- **Maintainability**: 6/10 - Function length and duplication hurt long-term maintenance
- **Code Quality**: 7/10 - Good practices overall, but needs refactoring
- **Test Coverage**: 9/10 - Excellent addition of positive and negative scenarios

### 💡 Recommendations

1. Extract helper functions from `run()` to improve testability
2. Create `verify_negative_scenario()` helper to eliminate duplication
3. Add try-except around type validation or document exception handling
4. Move log rotation outside the scenario loop
5. Define timeout constants with explanatory comments

**12 detailed review issues have been added to the Bob Findings panel for your review.**

File: tests/rados/test_osd_ok_to_upgrade_holistic_tests.py (hunk 2 above)
Comment (s-vipin) [reply tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:532]: This is in finally block. This would be required

**Comment by s-vipin** [comment]
## Commit Review Rating: **7.5/10** ⭐⭐⭐⭐⭐⭐⭐½

### Summary
Commit `47e65c8` "[osd ok-to-upgrade] stabilise the upgrade suite for osd ok-to-upgrade tests" introduces significant improvements to the OSD upgrade testing framework with better test coverage, improved constants management, and enhanced validation logic.

### ✅ Key Improvements Over Previous Version

**New Strengths:**
1. **Constants Defined**: Added `CLEAN_PG_TIMEOUT = 12000` and `MGR_DEBUG_LEVEL = "20/20"` - addresses the magic numbers issue
2. **Better Documentation**: Enhanced docstring with detailed scenario descriptions (lines 57-66)
3. **Improved Code Organization**: Constants at module level improve maintainability

### 📊 Detailed Analysis

**What Changed:**
- 2 files modified: test suite YAML + Python test implementation
- +238 lines, -72 lines (net +166 lines)
- Added 7 new test scenarios (3 positive, 4 negative)
- Introduced comprehensive OSD downgrade mechanism
- Enhanced logging and validation

**Functionality: 8.5/10**
- ✅ Comprehensive test coverage for bucket-type upgrades
- ✅ Proper negative test scenarios with error validation
- ✅ Robust cleanup via `_downgrade_osds_to_old_image()`
- ✅ Type safety with metadata validation
- ⚠️ Error handling still needs improvement (ValueError not caught)

**Code Quality: 7/10**
- ✅ Constants properly defined (CLEAN_PG_TIMEOUT, MGR_DEBUG_LEVEL)
- ✅ Fixed typo: `host_wihtout_osd_obj` → `host_without_osd_obj`
- ✅ Better command execution with `check_ec=True`
- ⚠️ Still has code duplication in negative scenarios (lines 320-400)
- ⚠️ `run()` function remains very long (600+ lines)

**Maintainability: 6.5/10**
- ✅ Constants improve readability
- ✅ Enhanced documentation in docstrings
- ⚠️ Function complexity unchanged - still needs refactoring
- ⚠️ Repeated patterns in negative test validation

**Test Coverage: 9/10**
- ✅ Excellent scenario coverage
- ✅ Both positive and negative test cases
- ✅ Validates both `ok-to-stop` and `ok-to-upgrade` behaviors
- ✅ Proper assertions for different upgrade paths

### 🎯 Specific Observations

**Positive Changes:**
1. **Line 27-28**: Constants defined - excellent improvement
2. **Line 57-66**: Comprehensive scenario documentation
3. **Line 407**: MGR_DEBUG_LEVEL constant used consistently
4. **Line 779**: CLEAN_PG_TIMEOUT constant used properly
5. **Line 739-742**: Type validation for metadata (good defensive programming)

**Remaining Issues:**
1. **Lines 320-400**: Four negative scenarios still follow identical patterns (DRY violation)
2. **Lines 61-529**: `run()` function still 468 lines - needs decomposition
3. **Line 739**: ValueError raised but not handled by caller
4. **Lines 465-495**: Log validation could be more robust
5. **Line 407**: Log rotation still inside scenario loop (performance concern)

### 💡 Recommendations for Next Iteration

**High Priority:**
1. Extract negative scenario validation into helper function:
   ```python
   def _verify_negative_scenario(rados_obj, cmd, expected_error_msg):
       out = list(rados_obj.client.exec_command(cmd=cmd))[1]
       if expected_error_msg in out:
           log.info(f"Expected failure: {out}")
           return True
       log.error(f"Unexpected output: {out}")
       return False
   ```

2. Wrap type validation in try-except or document exception handling requirement

**Medium Priority:**
3. Refactor `run()` into smaller functions (setup, execute, validate, cleanup)
4. Move log rotation outside scenario loop
5. Add retry logic to OSD downgrade operations

**Low Priority:**
6. Standardize comment formatting
7. Add more descriptive error messages

### 📈 Comparison to Previous Commit

| Aspect | Previous (75ac30d) | Current (47e65c8) | Change |
|--------|-------------------|-------------------|---------|
| Constants | ❌ Magic numbers | ✅ Named constants | +1.0 |
| Documentation | ⚠️ Basic | ✅ Comprehensive | +0.5 |
| Code Quality | 7/10 | 7/10 | 0 |
| Maintainability | 6/10 | 6.5/10 | +0.5 |
| **Overall** | **7.0/10** | **7.5/10** | **+0.5** |

### ✨ Summary

This commit shows **measurable improvement** with better constants management and documentation. The core functionality is solid with excellent test coverage. Main areas for improvement remain code organization (function length) and DRY principle violations in negative test scenarios. The +0.5 point improvement reflects the tangible enhancements in code maintainability through proper constant definitions.

**Recommendation**: Approve with suggestions for follow-up refactoring to address function complexity and code duplication.

**Comment by s-vipin** [comment]
## Commit Review: **8/10** ⭐⭐⭐⭐⭐⭐⭐⭐

### 📋 Commit Details
- **Hash**: 47e65c8 (or latest)
- **Message**: "[osd ok-to-upgrade] stabilise the upgrade suite for osd ok-to-upgrade tests"
- **Files Changed**: 2 (test suite YAML + Python test)
- **Lines**: +244, -78 (net +166)

---

## ✅ Strengths (What's Good)

### 1. **Excellent Test Coverage** (9/10)
- ✅ Added 7 new test scenarios (3 positive + 4 negative)
- ✅ Comprehensive bucket-type upgrade testing (rack, chassis, host)
- ✅ Proper negative test validation for parameter conflicts
- ✅ Both `ok-to-stop` and `ok-to-upgrade` command validation

### 2. **Code Quality Improvements** (8/10)
- ✅ **Constants defined**: `CLEAN_PG_TIMEOUT = 12000`, `MGR_DEBUG_LEVEL = "20/20"`
- ✅ **Bug fix**: `host_wihtout_osd_obj` → `host_without_osd_obj`
- ✅ **Better command execution**: Added `check_ec=True` for error checking
- ✅ **Enhanced logging**: Debug level configuration and log rotation
- ✅ **Type safety**: Added metadata type validation with proper error logging

### 3. **Robust Cleanup** (8/10)
- ✅ New `_downgrade_osds_to_old_image()` function for proper test cleanup
- ✅ Removes and re-adds OSDs to ensure clean state
- ✅ Waits for PG sets to be clean between operations
- ✅ **Documented exception**: Added docstring noting ValueError can be raised

### 4. **Better Documentation** (8/10)
- ✅ Comprehensive scenario descriptions in docstring (lines 57-66)
- ✅ Clear comments explaining negative test behavior
- ✅ Function docstring documents the ValueError exception

---

## ⚠️ Areas for Improvement

### 1. **Code Organization** (6/10)
**Issue**: The `run()` function is still 468 lines long
- Handles setup, execution, validation, and cleanup
- Difficult to test individual components
- Hard to maintain and understand flow

**Recommendation**: Refactor into smaller functions:
```python
def _setup_test_environment(...)
def _execute_upgrade_scenario(...)
def _validate_upgrade_results(...)
def _cleanup_test_environment(...)
```

### 2. **Code Duplication** (6/10)
**Issue**: Four negative scenarios (lines 320-400) follow identical patterns
- Same structure: build command → execute → check output → log
- Violates DRY principle
- Makes maintenance harder

**Recommendation**: Extract helper function:
```python
def _verify_negative_scenario(rados_obj, cmd, expected_error):
    out = list(rados_obj.client.exec_command(cmd=cmd))[1]
    if expected_error in out:
        log.info(f"Expected failure: {out}")
        return True
    log.error(f"Unexpected output: {out}")
    return False
```

### 3. **Exception Handling** (7/10)
**Issue**: ValueError is documented but not handled by caller
- Function raises ValueError (line 756-758)
- Called from finally block (line 541) without try-except
- Could interrupt cleanup if metadata is unexpected type

**Recommendation**: Add try-except in caller:
```python
try:
    _downgrade_osds_to_old_image(...)
except ValueError as e:
    log.error(f"Downgrade failed: {e}")
```

### 4. **Minor Issues** (7/10)
- Log rotation happens inside scenario loop (performance concern)
- Inconsistent comment formatting throughout
- grep operations don't distinguish "no matches" vs "command failed"

---

## 📊 Scoring Breakdown

| Category | Score | Weight | Notes |
|----------|-------|--------|-------|
| **Functionality** | 9/10 | 30% | Excellent test coverage, robust features |
| **Code Quality** | 8/10 | 25% | Good improvements, constants defined |
| **Maintainability** | 6/10 | 25% | Function length and duplication issues |
| **Documentation** | 8/10 | 10% | Good docstrings and comments |
| **Error Handling** | 7/10 | 10% | Documented but not fully handled |

**Weighted Score**: (9×0.3) + (8×0.25) + (6×0.25) + (8×0.1) + (7×0.1) = **7.8/10** → **8/10**

---

## 🎯 Key Improvements from Previous Versions

1. ✅ **Constants added** - No more magic numbers
2. ✅ **Exception documented** - ValueError now in docstring  
3. ✅ **Better logging** - Error logged before raising exception
4. ✅ **Enhanced docs** - Comprehensive scenario descriptions

---

## 💡 Recommendations for Next Iteration

### High Priority
1. **Refactor `run()` function** - Break into smaller, testable functions
2. **Extract negative test helper** - Eliminate code duplication
3. **Add exception handling** - Wrap `_downgrade_osds_to_old_image()` call

### Medium Priority
4. **Move log rotation** - Outside scenario loop for better performance
5. **Improve grep validation** - Distinguish between no matches vs errors

### Low Priority
6. **Standardize comments** - Consistent formatting throughout
7. **Add more assertions** - Validate intermediate states

---

## ✨ Final Verdict

**Strong commit with excellent test coverage and meaningful improvements.** The addition of constants, bug fixes, and comprehensive testing scenarios significantly enhance the codebase. Main areas for improvement are code organization (function length) and DRY principle violations. The documented ValueError shows good awareness, though full exception handling would make it production-ready.

**Recommendation**: ✅ **Approve** with suggestions for follow-up refactoring.

**Comment by s-vipin** [comment]
# Commit Review & Rating

## 🎯 Overall Score: **8.5/10** ⭐⭐⭐⭐⭐⭐⭐⭐½

**Commit**: `[osd ok-to-upgrade] stabilise the upgrade suite for osd ok-to-upgrade tests`

---

## 📊 Summary Statistics
- **Files Modified**: 2
- **Lines Added**: 250
- **Lines Removed**: 82
- **Net Change**: +168 lines
- **New Test Scenarios**: 7 (3 positive, 4 negative)
- **New Functions**: 1 (`_downgrade_osds_to_old_image`)

---

## ✅ Major Strengths

### 1. **Exceptional Test Coverage** (9.5/10)
- ✅ Added comprehensive bucket-type upgrade scenarios (rack, chassis, host)
- ✅ Four negative test cases validate parameter conflicts
- ✅ Tests both `ok-to-stop` and `ok-to-upgrade` command paths
- ✅ Validates log output to ensure correct upgrade strategy used

### 2. **Excellent Code Quality** (9/10)
- ✅ **Constants defined**: `CLEAN_PG_TIMEOUT = 12000`, `MGR_DEBUG_LEVEL = "20/20"`
- ✅ **Bug fixed**: Typo `host_wihtout_osd_obj` → `host_without_osd_obj`
- ✅ **Better error handling**: Added `check_ec=True` for command execution
- ✅ **Type safety**: Validates metadata is a list before processing
- ✅ **Exception handling**: **NEW!** Added try-except around downgrade function (line 541-546)

### 3. **Robust Cleanup Mechanism** (9/10)
- ✅ New `_downgrade_osds_to_old_image()` function ensures clean test state
- ✅ Properly removes and re-adds OSDs with old image
- ✅ Waits for clean PG sets between operations
- ✅ Handles edge case where no OSDs need downgrade

### 4. **Professional Documentation** (8.5/10)
- ✅ Comprehensive docstring with all scenario descriptions
- ✅ Function docstring documents ValueError exception
- ✅ Clear comments explaining negative test behavior
- ✅ Inline comments for complex logic

### 5. **Production-Ready Error Handling** (9/10) 🆕
- ✅ **ValueError now caught** in finally block with try-except
- ✅ Logs error but allows cleanup to continue
- ✅ Prevents test framework crashes from unexpected data

---

## ⚠️ Minor Areas for Improvement

### 1. **Function Length** (6/10)
**Issue**: `run()` function is 468 lines
- Handles multiple responsibilities
- Could benefit from extraction into smaller functions

**Impact**: Medium - Makes testing and maintenance harder

### 2. **Code Duplication** (6.5/10)
**Issue**: Four negative scenarios follow identical pattern (lines 320-400)
- Same structure repeated 4 times
- Could be extracted into helper function

**Impact**: Low-Medium - Maintenance burden if error handling changes

### 3. **Performance Consideration** (7/10)
**Issue**: Log rotation happens inside scenario loop (line 407-412)
- Rotates logs for every scenario
- Could be moved outside loop

**Impact**: Low - Minor performance overhead

---

## 📈 Detailed Scoring

| Category | Score | Weight | Weighted |
|----------|-------|--------|----------|
| **Test Coverage** | 9.5 | 30% | 2.85 |
| **Code Quality** | 9.0 | 25% | 2.25 |
| **Error Handling** | 9.0 | 20% | 1.80 |
| **Documentation** | 8.5 | 15% | 1.28 |
| **Maintainability** | 6.5 | 10% | 0.65 |
| **TOTAL** | | **100%** | **8.83** |

**Rounded Score**: **8.5/10**

---

## 🎉 Key Improvements in This Version

Compared to earlier iterations, this commit now includes:

1. ✅ **Exception handling added** - ValueError caught in finally block
2. ✅ **Constants properly defined** - No magic numbers
3. ✅ **Comprehensive documentation** - All scenarios documented
4. ✅ **Type validation** - Defensive programming with metadata check
5. ✅ **Better logging** - Debug level configuration for troubleshooting

---

## 💡 Recommendations (Optional Enhancements)

### For Future Iterations:

**Priority 1 - Refactoring** (Would raise score to 9/10)
```python
# Extract helper for negative tests
def _verify_negative_upgrade_scenario(rados_obj, cmd, expected_error):
    """Verify upgrade command fails with expected error."""
    out = list(rados_obj.client.exec_command(cmd=cmd))[1]
    if expected_error in out:
        log.info(f"Expected failure: {out}")
        return True
    log.error(f"Unexpected output: {out}")
    return False
```

**Priority 2 - Organization** (Would raise score to 9.5/10)
```python
# Break run() into logical sections
def _setup_upgrade_test(...)
def _execute_upgrade_scenario(...)
def _validate_upgrade_results(...)
def _cleanup_upgrade_test(...)
```

**Priority 3 - Performance**
- Move log rotation outside scenario loop
- Only rotate once per test run

---

## ✨ Final Verdict

**This is an excellent commit that significantly improves the test suite.** The addition of exception handling in the finally block addresses the critical issue from previous versions. Combined with proper constants, comprehensive test coverage, and robust cleanup mechanisms, this commit demonstrates professional software engineering practices.

### Strengths Summary:
- ✅ Production-ready error handling
- ✅ Comprehensive test scenarios
- ✅ Clean, well-documented code
- ✅ Proper cleanup mechanisms
- ✅ Type safety and validation

### Minor Improvements Possible:
- Function length (cosmetic issue)
- Code duplication (low priority)
- Performance optimization (negligible impact)

**Recommendation**: ✅ **APPROVE** - Ready for merge. Optional refactoring can be done in follow-up commits.

---

## 🏆 Rating Breakdown

- **8.5/10** = **Excellent** commit
- Would be 9/10 with function extraction
- Would be 9.5/10 with full refactoring
- Current state is production-ready and merge-worthy

**Great work on addressing the error handling!** 🎉

File: tests/rados/test_osd_ok_to_upgrade_holistic_tests.py [hunk 3]
```diff
@@ -55,6 +73,7 @@ def run(ceph_cluster, **kw):
       ceph orch upgrade resume.
     """
     config = kw["config"]
+    log.info(f"config: {config}")
```
Comment (harshkumarRH) [inline tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:75]: Why do we need this?

**Review by harshkumarRH — DISMISSED** [review]
Looks good

File: tests/rados/test_osd_ok_to_upgrade_holistic_tests.py (hunk 3 above)
Comment (s-vipin) [reply tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:75]: Had added for debugging purposes.

**Review by pdhiran — DISMISSED** [review]
looks good

**Review by harshkumarRH — APPROVED** [review]
Looks good

**Review by pdhiran — APPROVED** [review]
looks good
